### PR TITLE
New Adapter and Engine methods templates

### PR DESCRIPTION
**Pull request by alelom**
You can test the **Toolkit template** by placing the .zip files in:
C:\userName\Documents\Visual Studio 20xx\Templates\ProjectTemplates

You can test the **Engine methods templates** by placing the non-zipped folders containing them in:
C:\userName\Documents\Visual Studio 20xx\Templates\ItemTemplates

Changelog:
- Updated Adapter Template
- Updated Engine methods Template

### CODE REVIEW

**Comment by enarhi** [comment]
The current toolkit template does not build ootb. The highlighted text in the image below causes the issue, looks like it should not be there or needs to be reconfigured.

image

**Comment by alelom** [comment]
> The current toolkit template does not build ootb. The highlighted text in the image below causes the issue, looks like it should not be there or needs to be reconfigured.
> 
> image

Thanks @enarhi , just spotted that 😄 corrected.

**Comment by kThorsager** [comment]
We should remove un-needed references:

- Common_Engine
- Geometry_oM
- Geometry_Engine
- Physical_oM

and mainly keep the essential ones:

- BHoM
- Reflection_oM
- Reflection_Engine

to avoid reference bloat

File: templates/Engine method templates/ModifyTemplate/ModifyTemplate.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
+using BH.oM.Base;
+using BH.oM.Reflection.
```
Comment (kThorsager) [inline templates/Engine method templates/ModifyTemplate/ModifyTemplate.cs]: The `.` gives an error

File: templates/Engine method templates/QueryTemplate/QueryTemplate.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2019, the respective contributors. All rights reserved.
```
Comment (kThorsager) [inline templates/Engine method templates/QueryTemplate/QueryTemplate.cs]: Copyright 2020

**Comment by kThorsager** [comment]
Like this and would love it if it also included a link to the wiki: https://github.com/BHoM/documentation/wiki/BHoM_Engine-Namespaces
image

**Comment by alelom** [comment]
> Like this and would love it if it also included a link to the wiki: https://github.com/BHoM/documentation/wiki/BHoM_Engine-Namespaces

I've reviewed what now is https://github.com/BHoM/documentation/wiki/BHoM_Engine-Classes.
Thinking twice, I considered that I'd prefer not to include any Wiki URL in the template, because I'd like to separate them, especially considering that a renaming of a Wiki page should not affect the code.

File: templates/Toolkit template/SoftwareName_Toolkit/SoftwareName_Adapter/CRUD/Update/_IUpdate.cs [hunk 3]
```diff
@@ -25,24 +25,25 @@
...
-        }
+		// Unlike the Create, Delete and Read, this method already exposes a simple implementation: it calls Delete and then Create.
+		// It can be overridden here keeping in mind the following:
+        // - it gets called once per each Type, and if equal objects are found;
+		// - the object equality is tested through this.AdapterComparers, that need to be implemented for each type.
+		// See the wiki for more info.
+		
+        // protected override bool IUpdate<T>(IEnumerable<T> objects, ActionConfig actionConfig = null) 
+        // {
+            // return base.UpdateObjects<T>(objects);
```
Comment (kThorsager) [inline templates/Toolkit template/SoftwareName_Toolkit/SoftwareName_Adapter/CRUD/Update/_IUpdate.cs]: Method does not exist if one un-comments it.
(suspect it should be base.IUpdate<T>...)

File: templates/Toolkit template/SoftwareName_Toolkit/SoftwareName_Adapter/Convert/ToBHoM/Node.cs [hunk 4]
```diff
@@ -25,20 +25,18 @@
...
         /**** Public Methods                            ****/
         /***************************************************/
 
-        //Add methods for converting to BHoM from the specific software types. 
-        //Only do this if possible to do without any com-calls or similar to the adapter
-        //Example:
-        //public static Node ToBHoM(this $ext_safeprojectname$.Node node)
+        // Add methods for converting From BHoM to the specific software types
+		// Example:
+        //public static BH.oM.External.$ext_safeprojectname$.Node To$ext_safeprojectname$(this BH.oM.Structure.Elements.Node node)
```
[the remark this answers is not in the document]
Comment (FraserGreenroyd) [reply templates/Toolkit template/SoftwareName_Toolkit/SoftwareName_Adapter/Convert/ToBHoM/Node.cs]: Should be `FromSoftwareName` - `ToBHoM` and `FromBHoM` is not compliant at all.

File: templates/Toolkit template/SoftwareName_Toolkit/SoftwareName_Adapter/Convert/FromSoftwareName/Node.cs [hunk 5]
```diff
@@ -25,19 +25,18 @@
...
         /***************************************************/
         /**** Public Methods                            ****/
         /***************************************************/
 
-        //Add methods for converting From BHoM to the specific software types, if possible to do without any BHoM calls
-        //Example:
-        //public static $ext_safeprojectname$.Node To$ext_safeprojectname$(this Node node)
+        // Add methods for converting to BHoM from the specific software types. 
+        // Example:
+        //public static BH.oM.Structure.Elements.Node ToBHoM(this BH.oM.External.$ext_safeprojectname$.Node node)
```
Comment (kThorsager) [inline templates/Toolkit template/SoftwareName_Toolkit/SoftwareName_Adapter/Convert/FromSoftwareName/Node.cs]: `From$ext_safeprojectname$`

**Comment by kThorsager** [comment]
@alelom I'm still getting issues I've seen that you have fixed in the files for the engine methods, is there something similar to the zipping that you need to do for them?

File: templates/Engine method templates/CreateTemplate/CreateTemplate.cs [hunk 6]
```diff
@@ -21,34 +21,39 @@
...
-using System.Threading.Tasks;
+using System.Collections.Generic;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
 using BH.oM.Base;
-using BH.oM.Structure.Elements;
-using BH.oM.Structure.Constraints;
 
-namespace BH.Adapter.$ext_safeprojectname$
+namespace BH.Engine.External.SoftwareName // Replace SoftwareName with the name of your Toolkit software.
```
Comment (IsakNaslundBh) [inline templates/Engine method templates/CreateTemplate/CreateTemplate.cs]: Namespace as above.

`namespace $rootnamespace$` with comment about removing 'Create' if it exists I think.

File: templates/Engine method templates/ModifyTemplate/ModifyTemplate.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
+using BH.oM.Base;
+using BH.oM.Reflection;
+using BH.Engine.Base;
+
+namespace BH.Engine.External.SoftwareName // Replace SoftwareName with the name of your Toolkit software.
```
Comment (IsakNaslundBh) [inline templates/Engine method templates/ModifyTemplate/ModifyTemplate.cs]: Namespace as above.

`namespace $rootnamespace$` with comment about removing 'Create' if it exists I think.

File: templates/Engine method templates/ModifyTemplate/ModifyTemplate.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+{
+    public static partial class Modify
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+		// All public engine methods are reflected into the UIs.
```
Comment (IsakNaslundBh) [inline templates/Engine method templates/ModifyTemplate/ModifyTemplate.cs]: Same here about comments. Really good, as long as they do not start flooding. And again, might not be a superbig issue if they do. Mostly about consistency 😃

File: templates/Engine method templates/ModifyTemplate/ModifyTemplate.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        {
+			// // - First thing, clone the object to ensure immutability.
+			// // - You can do a light Shallow clone if the object type is BHoMObject and a shallow clone is enough for your case:
+			// SomeObjectType objClone = obj.GetShallowClone() as SomeObjectType;
+			
+			// // - Otherwise, BH.Engine.Base offers an efficient DeepClone method.
+			// SomeObjectType objClone = BH.Engine.Base.Query.DeepClone(obj);
+			
+			// If you need to log Error or Warnings and expose them to the user, use:
+			// BH.Engine.Reflection.Compute.RecordError("Error text") or RecordWarning
```
Comment (IsakNaslundBh) [inline templates/Engine method templates/ModifyTemplate/ModifyTemplate.cs]: 👍

File: templates/Engine method templates/ComputeTemplate/ComputeTemplate.cs [hunk 10]
```diff
@@ -21,34 +21,41 @@
...
         /***************************************************/
+		
+		/***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+		// Include here any private method that the public one might need to reference.
+		// Private methods are not reflected in the UIs.
+		
+		
+		/***************************************************/
```
Comment (IsakNaslundBh) [inline templates/Engine method templates/ComputeTemplate/ComputeTemplate.cs]: little bit nervous we will end up with this in many files where not needed, but a tradeoff I guess

File: templates/Engine method templates/QueryTemplate/QueryTemplate.cs [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+ */
+
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
+using BH.oM.Base;
+
+namespace BH.Engine.External.SoftwareName // Replace SoftwareName with the name of your Toolkit software.
```
Comment (IsakNaslundBh) [inline templates/Engine method templates/QueryTemplate/QueryTemplate.cs]: Namespace as above.

`namespace $rootnamespace$` with comment about removing 'Query' if it exists I think.

File: templates/Engine method templates/ComputeTemplate/ComputeTemplate.cs (hunk 10 above)
Comment (alelom) [reply templates/Engine method templates/ComputeTemplate/ComputeTemplate.cs]: Removed

File: templates/Engine method templates/QueryTemplate/QueryTemplate.cs (hunk 11 above)
Comment (alelom) [reply templates/Engine method templates/QueryTemplate/QueryTemplate.cs]: Done

**Comment by kThorsager** [comment]
honestly not sold on this
image

It could at least not write BH.Engine twice?

also where is the `Convert` template?

**Comment by alelom** [comment]
> honestly not sold on this
😅 Sorry! Modifying it now

> also where is the `Convert` template?
There never was a Convert template, and I didn't add one since it's moved in the Adapter project now. I guess there are still use cases where you might want a convert in the Engine, but since I wouldn't suggest that as a standard practice, I'd avoid creating a template for it.

**Comment by alelom** [comment]
@IsakNaslundBh @kThorsager could you please re-review?

**Comment by alelom** [comment]
> LGTM, just a minor thing I'm prepared to let go as I doubt that it's your fault (might be my old Visual Studio)

Good spot - did that for me too. I swear VS templates are its most annoying part. Shortened the description so now it's not trimmed anymore.

**Comment by al-fisher** [comment]
Thanks @IsakNaslundBh @enarhi @kThorsager @alelom 
Will merge this now for use. If any of you have any future tweaks can raise as issues 👍

**Pull request merged**